### ISSUE TITLE
Integrating fluentvalidator throw an error on input fields

### ISSUE THREAD

**Issue by arivera12** [chanan/BlazorStrap#278]
Hi,

I was working integrating `FluentValidation ` using Steve Sanderson integration approach.

https://blog.stevensanderson.com/2019/09/04/blazor-fluentvalidation/

https://gist.github.com/SteveSandersonMS/090145d7511c5190f62a409752c60d00#file-fluentvalidator-cs

But the input on change event the value is binded correctly but the validator throws an error `onblur`.
![image](https://user-images.githubusercontent.com/5600102/77859543-1a84b200-71d8-11ea-8ff8-6e451271cc12.png)

The problem seems to occur onblur event when tries to validate exactly when tries to cast to `MulticastDelegate`.
![image](https://user-images.githubusercontent.com/5600102/77859610-5e77b700-71d8-11ea-86b5-7b8bf10f8110.png)

This is my class implementation with some edits from the original post from Steve Sanderson.
```
using FluentValidation;
using Microsoft.AspNetCore.Components;
using Microsoft.AspNetCore.Components.Forms;
using System;
using System.Linq;

namespace Extensions
{
    public class FluentValidator<TValidator> : ComponentBase where TValidator : IValidator, new()
    {
        private readonly static char[] separators = new[] { '.', '[' };
        private TValidator validator;
        [CascadingParameter] private EditContext EditContext { get; set; }
        protected override void OnInitialized()
        {
            validator = new TValidator();
            var messages = new ValidationMessageStore(EditContext);
            EditContext.OnValidationRequested += (sender, eventArgs)
                => ValidateModel((EditContext)sender, messages);

            EditContext.OnFieldChanged += (sender, eventArgs)
                => ValidateModelField((EditContext)sender, messages, eventArgs);  
        }
        private void ValidateModel(EditContext editContext, ValidationMessageStore messages)
        {
            var validationResult = validator.Validate(editContext.Model);
            messages.Clear();
            foreach (var error in validationResult.Errors)
            {
                var fieldIdentifier = ToFieldIdentifier(editContext, error.PropertyName);
                messages.Add(fieldIdentifier, OL.Resourcer.GetString(error.ErrorMessage));
            }
            editContext.NotifyValidationStateChanged();
        }
        private void ValidateModelField(EditContext editContext, ValidationMessageStore messages, FieldChangedEventArgs fieldChangedEventArgs)
        {
            var validationResult = validator.Validate(editContext.Model);
            messages.Clear();
            foreach (var error in validationResult.Errors.Where(w => w.PropertyName == fieldChangedEventArgs.FieldIdentifier.FieldName))
            {
                var fieldIdentifier = ToFieldIdentifier(editContext, error.PropertyName);
                messages.Add(fieldIdentifier, OL.Resourcer.GetString(error.ErrorMessage));
            }
            editContext.NotifyValidationStateChanged();
        }
        private static FieldIdentifier ToFieldIdentifier(EditContext editContext, string propertyPath)
        {
            // This method parses property paths like 'SomeProp.MyCollection[123].ChildProp'
            // and returns a FieldIdentifier which is an (instance, propName) pair. For example,
            // it would return the pair (SomeProp.MyCollection[123], "ChildProp"). It traverses
            // as far into the propertyPath as it can go until it finds any null instance.

            var obj = editContext.Model;

            while (true)
            {
                var nextTokenEnd = propertyPath.IndexOfAny(separators);
                if (nextTokenEnd < 0)
                {
                    return new FieldIdentifier(obj, propertyPath);
                }

                var nextToken = propertyPath.Substring(0, nextTokenEnd);
                propertyPath = propertyPath.Substring(nextTokenEnd + 1);

                object newObj;
                if (nextToken.EndsWith("]"))
                {
                    // It's an indexer
                    // This code assumes C# conventions (one indexer named Item with one param)
                    nextToken = nextToken.Substring(0, nextToken.Length - 1);
                    var prop = obj.GetType().GetProperty("Item");
                    var indexerType = prop.GetIndexParameters()[0].ParameterType;
                    var indexerValue = Convert.ChangeType(nextToken, indexerType);
                    newObj = prop.GetValue(obj, new object[] { indexerValue });
                }
                else
                {
                    // It's a regular property
                    var prop = obj.GetType().GetProperty(nextToken);
                    if (prop == null)
                    {
                        throw new InvalidOperationException($"Could not find property named {nextToken} on object of type {obj.GetType().FullName}.");
                    }
                    newObj = prop.GetValue(obj);
                }

                if (newObj == null)
                {
                    // This is as far as we can go
                    return new FieldIdentifier(obj, nextToken);
                }

                obj = newObj;
            }
        }
    }
}

```

Then inside a `<BSForm>`

`<FluentValidator TValidator="MyValidatorClass"></FluentValidator>`

Not sure if this should be considered a bug or a enhancement feature but for me seems a bug cause we are only changing the default validator to use fluentvalidation instead of DataAnnotations.

**Comment by chanan** [comment]
I havent used the validators in forms yet, I will take a look at it this coming weekend

**Comment by arivera12** [comment]
This will be a huge improvement if its fixed and supported. FluentValidation is more powerful than .net DataAnnotations validations system.

**Comment by chucker** [comment]
I'm not sure why the code uses reflection to get `OnFieldChanged` at all, since it's public. Maybe it wasn't before? Anywhere, it seems like calling [`NotifyFieldChanged`](https://github.com/dotnet/aspnetcore/blob/9a792f2a270dc68bac47f65cccf6ac56f6a062d5/src/Components/Forms/src/EditContext.cs#L69) is more appropriate here.

**Comment by arivera12** [comment]
@chucker It's this gonna get fixed?

**Comment by jbomhold3** [comment]
I'll work on this one. I have a working version of FluentValidation however what you posted should work as well. @chanan  we probably should add support out of the box for FluentValidation.

Heres how I did it. 

<FluentValidationValidator />
```C#
public static class EditContextFluentValidationExtensions
{
    public static EditContext AddFluentValidation(this EditContext editContext)
    {
        if (editContext == null)
        {
            throw new ArgumentNullException(nameof(editContext));
        }

        var messages = new ValidationMessageStore(editContext);

        editContext.OnValidationRequested +=
            (sender, eventArgs) => ValidateModel((EditContext)sender, messages);

        editContext.OnFieldChanged +=
            (sender, eventArgs) => ValidateField(editContext, messages, eventArgs.FieldIdentifier);

        return editContext;
    }

    private static void ValidateModel(EditContext editContext, ValidationMessageStore messages)
    {
        try
        {
            IValidator validator = GetValidatorForModel(editContext.Model);
            FluentValidation.Results.ValidationResult validationResults = validator.Validate(editContext.Model);

            messages.Clear();
            foreach (FluentValidation.Results.ValidationFailure validationResult in validationResults.Errors)
            {
                messages.Add(editContext.Field(validationResult.PropertyName), validationResult.ErrorMessage);
            }

            editContext.NotifyValidationStateChanged();
        }
        catch
        {
        }
    }

    private static void ValidateField(EditContext editContext, ValidationMessageStore messages, in FieldIdentifier fieldIdentifier)
    {
        try
        {
            var properties = new[] { fieldIdentifier.FieldName };
            var context = new ValidationContext(fieldIdentifier.Model, new PropertyChain(), new MemberNameValidatorSelector(properties));

            IValidator validator = GetValidatorForModel(fieldIdentifier.Model);
            FluentValidation.Results.ValidationResult validationResults = validator.Validate(context);

            messages.Clear(fieldIdentifier);
            foreach (var error in validationResults.Errors.Select(error => error.ErrorMessage))
            {
                messages.Add(fieldIdentifier, error);
            }

            editContext.NotifyValidationStateChanged();
        }
        catch
        {
        }
    }

    private static IValidator GetValidatorForModel(object model)
    {
        try
        {
            Type abstractValidatorType = typeof(AbstractValidator<>).MakeGenericType(model.GetType());
            Type modelValidatorType = model.GetType().Assembly.GetTypes().FirstOrDefault(t => t.IsSubclassOf(abstractValidatorType));
            var modelValidatorInstance = (IValidator)Activator.CreateInstance(modelValidatorType);
            return modelValidatorInstance;
        }
        catch
        {
            return default;
        }
    }
```

**Comment by jbomhold3** [comment]
Support added will be in the next build.
Add @using BlazorStrap.Components.Forms.Validators
<FluentValidator TValidator="CustomerValidator" />

**Comment by chanan** [comment]
Sorry can’t find the comment I just saw from @jbomhold3 So replying here

Don’t we already have a Blazor strap extensions repo that we never used?

**Comment by chanan** [comment]
@jbomhold3 why can’t I find your comments?

Eating dinner right now. I will try to join in about an hour. Otherwise if you are still using Skype I will find you there

**Comment by jbomhold3** [comment]
Moved and addressed in #286

**Comment by jbomhold3** [comment]
Extension Added for this
